go/database/mpt/proof.go

```diff
@@ -0,0 +1,197 @@
+// as they are present in the MPT path from root to the leaf to prove.
+func (p *nodeVisitor) Visit(node Node, _ NodeInfo) VisitResponse {
+	data := make([]byte, 0, 1024)
+	rlp, err := p.hasher.encode(node, p.nodeSource, data)
+	if err != nil {
+		p.err = err
+		return VisitResponseAbort
+	}
+	hash := common.Keccak256(rlp)
+
+	p.proof = append(p.proof, ProofElement{hash, rlp})
+
+	return VisitResponseContinue
+}
+
+// StorageProof represents a witness proof for a storage slot.
+type StorageProof proof
+
+// AccountProof represents a witness proof for an account.
+type AccountProof proof
+
+// ProofSource is a utility to create witness proofs for accounts and storage slots.
+// It has access to nodes of the MPT and a root.
+// The witness proof is created by iterating the MPT from the current root
+// down to a leaf node following the input address or  a storage key.
+// Each node encountered on the path is hashed and stored in a proof.
+// This proof represents the witness proof of the input address and storage key.
+// This proof is furthermore verified by iterating it as an MPT structure from the root down
+// to the leaf node while checking at every step that the encountered node is present in
+// the proof.
+type ProofSource struct {
+	nodeSource NodeSource
+	hasher     hasher
+	root       *NodeReference
+}
+
+// NewProofSource creates a new proof source with the input node source and root node.
+func NewProofSource(root *NodeReference, nodeSource NodeSource) *ProofSource {
+	return &ProofSource{
+		nodeSource: nodeSource,
+		hasher:     nodeSource.getConfig().Hashing.createHasher(),
+		root:       root,
+	}
+}
+
+// StorageProofs is an account and storage witness proof.
+// It contains the witness proof of the account and a list of storage proofs for the storage slots.
+// The witness proof is created by iterating the MPT from the root node
+// down to an account node following the input address.
+// Each node encountered on the path is hashed and stored in a proof.
+// Furthermore, the storage proofs are created by iterating the MPT from the account node
+// down to the leaf node following the storage key.
+// The proof contains one account and possibly multiple storage proofs.
+type StorageProofs struct {
+	accountProof AccountProof
+	storage      map[common.Key]StorageProof
+}
+
+// GetAccountProof returns the account proof.
+func (p *StorageProofs) GetAccountProof() AccountProof {
+	return p.accountProof
+}
+
+// GetStorageProof returns the storage proof for the input storage key.
+func (p *StorageProofs) GetStorageProof(key common.Key) StorageProof {
+	return p.storage[key]
+}
+
+// CreateAccountProof creates a witness proof for the account at the given address.
+// It iterates the MPT from the root down to the leaf node following the address.
+// Each node encountered on the path is hashed and stored in a proof.
+func (p *ProofSource) CreateAccountProof(address common.Address) (AccountProof, error) {
```

Comment:
The implementation of this function could call `CreateStorageProof` with an empty key list to avoid redundant code. Wdyt?